odl/space/cupy_tensors.py

```diff
@@ -0,0 +1,2819 @@
+
+        References
+        ----------
+        .. _corresponding NEP:
+           https://github.com/numpy/numpy/blob/master/doc/neps/\
+ufunc-overrides.rst
+
+        .. _interface documentation:
+           https://github.com/charris/numpy/blob/master/doc/source/reference/\
+arrays.classes.rst#special-attributes-and-methods
+
+        .. _general documentation on Numpy ufuncs:
+           https://docs.scipy.org/doc/numpy/reference/ufuncs.html
+
+        .. _reduceat documentation:
+           https://docs.scipy.org/doc/numpy/reference/generated/\
+numpy.ufunc.reduceat.html
+        """
+        # --- Process `out` --- #
+
+        # Unwrap out if provided. The output parameters are all wrapped
+        # in one tuple, even if there is only one.
+        out_tuple = kwargs.pop('out', ())
+
+        # Check number of `out` args, depending on `method`
+        if method == '__call__' and len(out_tuple) not in (0, ufunc.nout):
+            raise ValueError(
+                "need 0 or {} `out` arguments for `method='__call__'`, "
+                'got {}'.format(ufunc.nout, len(out_tuple)))
+        elif method != '__call__' and len(out_tuple) not in (0, 1):
+            raise ValueError(
+                "need 0 or 1 `out` arguments for `method={!r}`, "
+                'got {}'.format(method, len(out_tuple)))
+
+        # Catch wrong number of inputs
+        if method == '__call__' and len(inputs) != ufunc.nin:
+            raise ValueError(
+                "need {} inputs for `method='__call__'`, got {}"
+                ''.format(ufunc.nin, len(inputs)))
+
+        # We allow our own tensors, the data container type and
+        # `numpy.ndarray` objects as `out` (see docs for reason for the
+        # latter)
+        valid_types = (type(self), type(self.data), np.ndarray)
+        if not all(isinstance(o, valid_types) or o is None
+                   for o in out_tuple):
+            return NotImplemented
+
+        # Assign to `out` or `out1` and `out2`, respectively, unwrapping the
+        # data container
+        out = out1 = out2 = None
+        if len(out_tuple) == 1:
+            if isinstance(out_tuple[0], type(self)):
+                out = out_tuple[0].data
+            else:
+                out = out_tuple[0]
+        elif len(out_tuple) == 2:
+            if isinstance(out_tuple[0], type(self)):
+                out1 = out_tuple[0].data
+            else:
+                out1 = out_tuple[0]
+            if isinstance(out_tuple[1], type(self)):
+                out2 = out_tuple[1].data
+            else:
+                out2 = out_tuple[1]
+
+        # --- Process `inputs` --- #
+
+        # Pull out the data container of the inputs if necessary
+        inputs = [inp.data if isinstance(inp, type(self)) else inp
+                  for inp in inputs]
+
+        # Determine native ufunc vs. Numpy ufunc
+        if any(isinstance(o, np.ndarray) for o in out_tuple):
+            native_ufunc = None
```

Comment:
I need some input here @kohr-h, i should be able to merge these if statements, right?

Reply:
So by "native" I mean "CuPy" here, which could easily replace "native" in the names as well.

The reasoning for this distinction here is like this:
- If the output is (or contains) a Numpy array, use Numpy (=fallback). This is essential to support `numpy_array += odl_elem_using_cupy` for example, since it will use `out=(numpy_array,)`.
- If any of the operands is not either a scalar or a CuPy array, use Numpy. This is debatable, but any version will involve copies.
- If all inputs and all outputs are scalars or CuPy arrays, try to use the CuPy ufunc.

The success of "try to" depends on whether the function is implemented, which is true for almost all basic ufuncs, some reductions and accumulations, but nothing else. So there's another possibility for fallback here.

I guess you could merge the two first fallback cases logically.